gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -252,24 +256,50 @@ private WriteObjectResponse doResumableUpload() throws IOException {
 
       boolean objectFinalized = false;
       while (!objectFinalized) {
-        WriteObjectRequest insertRequest;
+        WriteObjectRequest insertRequest = null;
         if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
           insertRequest = buildRequestFromBufferedDataChunk(dataChunkMap, writeOffset);
           writeOffset += insertRequest.getChecksummedData().getContent().size();
         } else {
-          ByteString data =
-              ByteString.readFrom(
-                  ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE), MAX_BYTES_PER_MESSAGE);
-          dataChunkMap.put(writeOffset, data);
           if (dataChunkMap.size() >= channelOptions.getNumberOfBufferedRequests()) {
-            dataChunkMap.remove(dataChunkMap.firstKey());
+            /*
+             If dataChunkMap is full, get the committedWriteOffset. This will make a API call to the
+             server and add latency in this path/context. The upload is happening asynchronously in
+             a different context. So this API will not add latency to overall upload. It will only
+             throttle the upstream write call, which is fine.
+            */
+            committedWriteOffset = getCommittedWriteSizeWithRetries(uploadId);
+            logger.atFinest().log(
+                "Fetched committedWriteOffset: size:%d, numBuffers:%d, writeOffset:%d, committedWriteOffset:%d",
+                dataChunkMap.size(),
+                channelOptions.getNumberOfBufferedRequests(),
+                this.writeOffset,
+                committedWriteOffset);
+
+            // check and remove chunks from dataChunkMap
+            while (dataChunkMap.size() > 0 && dataChunkMap.firstKey() < committedWriteOffset) {
```

Comment:
Can we add additional tests to verify the change ?

Reply:
what happens if the commit offset >= first key , potentially with offset .

we can end up with uncontrolled buffer size